pkg/storegateway/bucket.go

```diff
@@ -611,6 +611,11 @@ func (s *BucketStore) Series(req *storepb.SeriesRequest, srv storegatewaypb.Stor
 	// but sometimes it can take minutes if the block isn't loaded and there is a surge in queries for unloaded blocks.
 	done, err := s.limitConcurrentQueries(ctx, stats)
 	if err != nil {
+		if errors.Is(err, errGateTimeout) {
+			// If the gate timed out, then we behave in the same way as if the blocks aren't discovered yet.
+			// The querier will try the blocks again on a different store-gateway replica.
```

Comment:
How is this situation logged by queriers? (I'm wondering if we'll get a misleading or confusing message logged by queriers along the lines of "store-gateway hasn't discovered blocks", when the truth is that the store-gateway timed out the request.)

Reply:
I have a similar concern here. Why not returning the error? The querier will retry the request on other store-gateways in case of error, unless it's detected as a non retriable error (but we make sure the error we return here is detected as retryable).

Reply:
> How is this situation logged by queriers? 

this is the SpanLog that the querier logs https://github.com/grafana/mimir/blob/fce7b3800b3839c9fd44b4ff0c510b96a0ba36cc/pkg/querier/blocks_store_queryable.go#L596

If all of the store-gateways cannot query a block, then this is the error that the querier generates: https://github.com/grafana/mimir/blob/fce7b3800b3839c9fd44b4ff0c510b96a0ba36cc/pkg/querier/blocks_store_queryable.go#L600-L601

https://github.com/grafana/mimir/blob/fce7b3800b3839c9fd44b4ff0c510b96a0ba36cc/pkg/querier/blocks_store_queryable.go#L596

---

> Why not returning the error? The querier will retry the request on other store-gateways in case of error, unless it's detected as a non retriable error (but we make sure the error we return here is detected as retryable).

That's not the case unfortunately. If the store-gateway returns an error, then the querier will immediately abort the query: 
https://github.com/grafana/mimir/blob/fce7b3800b3839c9fd44b4ff0c510b96a0ba36cc/pkg/querier/blocks_store_queryable.go#L569-L574

I did consider returning a special error, but that seemed more complicated because of response processing in [`fetchSeriesFromStores`](https://github.com/grafana/mimir/blob/fce7b3800b3839c9fd44b4ff0c510b96a0ba36cc/pkg/querier/blocks_store_queryable.go#L707). If you still think it's a better idea to have a special error and handle it properly, then I can do the necessary refactors; perhaps even adopt a similar error-handling pattern to the distributor-ingester ingestion error handling.

Reply:
> That's not the case unfortunately. If the store-gateway returns an error, then the querier will immediately abort the query:

I'm not sure about it. I think the "magic" is done by `queryF`. For example, look at `fetchSeriesFromStores()` implementation and its call to `shouldStopQueryFunc()`. I think the idea is that `queryF` only return a fatal error, while swallow errors that shouldn't stop the query execution.

Reply:
you're right there is `shouldStopQueryFunc`

https://github.com/grafana/mimir/blob/fce7b3800b3839c9fd44b4ff0c510b96a0ba36cc/pkg/querier/blocks_store_queryable.go#L755-L760

let me give it a try to add some error handling similar to distributor-ingester errors

Reply:
this ended up being more complicated than i expected too. I opened a PR to prepare for this change https://github.com/grafana/mimir/pull/7790. I could have copied some code too, but wanted to move towards reducing code duplication